File: repel-pointer-swift/Sources/MetalParticles/Renderer.swift
import Metal
import MetalKit
import simd
import Foundation

struct Particle {
    var position: SIMD2<Float>
    var velocity: SIMD2<Float>
    var originalPosition: SIMD2<Float>
}

struct Uniforms {
    var mousePosition: SIMD2<Float>
    var time: Float
    var resolution: SIMD2<Float>
    var repulsionRadius: Float
    var repulsionStrength: Float
}

class Renderer: NSObject, MTKViewDelegate {
    var device: MTLDevice!
    var commandQueue: MTLCommandQueue!
    var computePipelineState: MTLComputePipelineState!
    var renderPipelineState: MTLRenderPipelineState!
    
    var particleBuffer: MTLBuffer!
    var uniformsBuffer: MTLBuffer!
    
    var particles: [Particle] = []
    let particleCount = 5000
    
    var viewportSize: CGSize = .zero
    var mousePosition: CGPoint = .zero
    
    init?(metalKitView: MTKView) {
        super.init()
        
        guard let device = MTLCreateSystemDefaultDevice() else { return nil }
        self.device = device
        self.commandQueue = device.makeCommandQueue()
        
        metalKitView.device = device
        metalKitView.clearColor = MTLClearColor(red: 0.1, green: 0.1, blue: 0.1, alpha: 1.0)
        metalKitView.delegate = self
        
        buildPipelineStates()
        initParticles()
    }
    
    func buildPipelineStates() {
        var library: MTLLibrary?
        // Load the library
        do {
            // Try to load from default library first (if it exists)
            if let defaultLib = device.makeDefaultLibrary() {
                library = defaultLib
                print("Loaded default library")
            } else {
                // Fallback: Compile from source
                print("Default library not found, compiling from source...")
                guard let shaderPath = Bundle.module.path(forResource: "Shaders", ofType: "metal"),
                      let headerPath = Bundle.module.path(forResource: "ShaderDefinitions", ofType: "h") else {
                    print("Could not find Shaders.metal or ShaderDefinitions.h in bundle")
                    return
                }
                
                let shaderSource = try String(contentsOfFile: shaderPath, encoding: .utf8)
                let headerSource = try String(contentsOfFile: headerPath, encoding: .utf8)
                
                // Simple include replacement
                let finalSource = shaderSource.replacingOccurrences(of: "#include \"ShaderDefinitions.h\"", with: headerSource)
                
                library = try device.makeLibrary(source: finalSource, options: nil)
                print("Compiled library from source")
            }
        } catch {
            print("Library error: \(error)")
            return
        }
        
        guard let lib = library else {
            print("Could not create default library")
            return
        }
        print("Library loaded successfully")
        
        if !device.supportsFamily(.metal3) {
            print("Warning: Device does not support Metal 3")
        }
        
        // Compute Pipeline
        guard let kernelFunction = lib.makeFunction(name: "updateParticles") else {
            print("Could not create kernel function")
            return
        }
        
        let computeDescriptor = MTLComputePipelineDescriptor()
        computeDescriptor.computeFunction = kernelFunction
        computeDescriptor.label = "Particle Update"
        
        do {
            computePipelineState = try device.makeComputePipelineState(descriptor: computeDescriptor, options: [], reflection: nil)
        } catch {
            print("Compute pipeline error: \(error)")
        }
        
        // Render Pipeline
        let pipelineDescriptor = MTLRenderPipelineDescriptor()
        pipelineDescriptor.vertexFunction = lib.makeFunction(name: "vertexShader")
        pipelineDescriptor.fragmentFunction = lib.makeFunction(name: "fragmentShader")
        pipelineDescriptor.colorAttachments[0].pixelFormat = .bgra8Unorm
        
        do {
            renderPipelineState = try device.makeRenderPipelineState(descriptor: pipelineDescriptor)
        } catch {
            print("Render pipeline error: \(error)")
        }
    }
    
    func initParticles() {
        // Initial distribution will be handled in mtkView(_:drawableSizeWillChange:)
        // or we can just allocate here and set positions later.
        // Let's allocate buffer first.
        
        let bufferSize = MemoryLayout<Particle>.stride * particleCount
        particleBuffer = device.makeBuffer(length: bufferSize, options: .storageModeShared)
    }
    
    func resetParticles(size: CGSize) {
        guard let buffer = particleBuffer else { return }
        
        var pointer = buffer.contents().bindMemory(to: Particle.self, capacity: particleCount)
        
        let cols = Int(sqrt(Double(particleCount)))
        let rows = particleCount / cols
        
        let spacingX = Float(size.width) / Float(cols)
        let spacingY = Float(size.height) / Float(rows)
        
        for i in 0..<particleCount {
            let col = i % cols
            let row = i / cols
            
            let x = Float(col) * spacingX + spacingX * 0.5
            let y = Float(row) * spacingY + spacingY * 0.5
            
            let pos = SIMD2<Float>(x, y)
            
            pointer.pointee = Particle(position: pos, velocity: .zero, originalPosition: pos)
            pointer = pointer.advanced(by: 1)
        }
    }
    
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        viewportSize = size
        resetParticles(size: size)
    }
    
    func draw(in view: MTKView) {
        guard let commandBuffer = commandQueue.makeCommandBuffer(),
              let drawable = view.currentDrawable,
              let _ = view.currentRenderPassDescriptor else { return }
        
        // Update Uniforms
        var uniforms = Uniforms(
            mousePosition: SIMD2<Float>(Float(mousePosition.x), Float(mousePosition.y)),
            time: Float(CACurrentMediaTime()),
            resolution: SIMD2<Float>(Float(viewportSize.width), Float(viewportSize.height)),
            repulsionRadius: 300.0,
            repulsionStrength: 2.0
        )
        
        // Compute Pass
        if let computeEncoder = commandBuffer.makeComputeCommandEncoder() {
            if let pipeline = computePipelineState {
                computeEncoder.setComputePipelineState(pipeline)
                computeEncoder.setBuffer(particleBuffer, offset: 0, index: 0)
                computeEncoder.setBytes(&uniforms, length: MemoryLayout<Uniforms>.stride, index: 1)
                
                let threadGroupSize = MTLSize(width: pipeline.threadExecutionWidth, height: 1, depth: 1)
                let threadGroups = MTLSize(width: (particleCount + threadGroupSize.width - 1) / threadGroupSize.width, height: 1, depth: 1)
                
                computeEncoder.dispatchThreadgroups(threadGroups, threadsPerThreadgroup: threadGroupSize)
            }
            computeEncoder.endEncoding()
        }
        
        // Render Pass
        if let descriptor = view.currentRenderPassDescriptor,
           let renderEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: descriptor) {
            if let pipeline = renderPipelineState {
                renderEncoder.setRenderPipelineState(pipeline)
                renderEncoder.setVertexBuffer(particleBuffer, offset: 0, index: 0)
                renderEncoder.setVertexBytes(&uniforms, length: MemoryLayout<Uniforms>.stride, index: 1)
                
                renderEncoder.drawPrimitives(type: .point, vertexStart: 0, vertexCount: particleCount)
            }
            renderEncoder.endEncoding()
        }
        
        commandBuffer.present(drawable)
        commandBuffer.commit()
    }
    
    func updateMousePosition(_ point: CGPoint) {
        // Convert view coordinates to pixel coordinates if necessary,
        // but here we assume the point passed is already relative to the view.
        // We need to scale it to the drawable size (Retina display).
        
        // Note: The point passed from SwiftUI will be in points.
        // The shader expects pixels.
        // We'll handle scaling in the view.
        self.mousePosition = point
    }
}

File: repel-pointer-cpp/src/Renderer.cpp
#include "Renderer.hpp"

// Include metal-cpp headers only in the .cpp file
#include <Foundation/Foundation.hpp>
#include <Metal/Metal.hpp>
#include <QuartzCore/QuartzCore.hpp>

#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <unistd.h>

std::string Renderer::readFile(const std::string &filepath) {
  std::ifstream t(filepath);
  if (!t.is_open()) {
    std::cerr << "Failed to open " << filepath
              << ". Current working directory: " << getcwd(NULL, 0)
              << std::endl;
    return "";
  }
  std::stringstream buffer;
  buffer << t.rdbuf();
  return buffer.str();
}

Renderer::Renderer(CA::MetalLayer *layer) {
  _device = MTL::CreateSystemDefaultDevice();
  _layer = layer;

  if (_layer && _device) {
    _layer->setDevice(_device);
    _layer->setPixelFormat(MTL::PixelFormatBGRA8Unorm);
  }

  _commandQueue = _device->newCommandQueue();

  _viewportWidth = 800.0f;
  _viewportHeight = 600.0f;
  _mouseX = 0.0f;
  _mouseY = 0.0f;

  buildPipelineStates();
  initParticles();
  resetParticles(_viewportWidth, _viewportHeight);
}

Renderer::~Renderer() {
  if (_particleBuffer)
    _particleBuffer->release();
  if (_computePipelineState)
    _computePipelineState->release();
  if (_renderPipelineState)
    _renderPipelineState->release();
  if (_commandQueue)
    _commandQueue->release();
  if (_device)
    _device->release();
}

void Renderer::setLayer(CA::MetalLayer *layer) {
  _layer = layer;
  if (_layer && _device) {
    _layer->setDevice(_device);
    _layer->setPixelFormat(MTL::PixelFormatBGRA8Unorm);
  }
}

void Renderer::buildPipelineStates() {
  // Load shader source
  std::string shaderSource = readFile("src/Shaders.metal");
  std::string headerSource = readFile("src/ShaderDefinitions.h");

  if (shaderSource.empty() || headerSource.empty()) {
    std::cerr << "Could not load shader sources. Ensure you are running from "
                 "the project root."
              << std::endl;
    return;
  }

  // Manual #include replacement
  std::string search = "#include \"ShaderDefinitions.h\"";
  size_t pos = shaderSource.find(search);
  if (pos != std::string::npos) {
    shaderSource.replace(pos, search.length(), headerSource);
  }

  NS::Error *error = nullptr;
  NS::String *source =
      NS::String::string(shaderSource.c_str(), NS::UTF8StringEncoding);
  MTL::Library *library = _device->newLibrary(source, nullptr, &error);

  if (!library) {
    std::cerr << "Library compile error: "
              << (error ? error->localizedDescription()->utf8String()
                        : "unknown")
              << std::endl;
    return;
  }

  // Compute Pipeline
  NS::String *kernelName =
      NS::String::string("updateParticles", NS::UTF8StringEncoding);
  MTL::Function *kernelFunction = library->newFunction(kernelName);
  if (!kernelFunction) {
    std::cerr << "Could not find kernel function" << std::endl;
    library->release();
    return;
  }

  _computePipelineState =
      _device->newComputePipelineState(kernelFunction, &error);

  if (!_computePipelineState) {
    std::cerr << "Compute pipeline error: "
              << (error ? error->localizedDescription()->utf8String()
                        : "unknown")
              << std::endl;
  }
  kernelFunction->release();

  // Render Pipeline
  MTL::RenderPipelineDescriptor *renderDescriptor =
      MTL::RenderPipelineDescriptor::alloc()->init();

  NS::String *vertexName =
      NS::String::string("vertexShader", NS::UTF8StringEncoding);
  NS::String *fragmentName =
      NS::String::string("fragmentShader", NS::UTF8StringEncoding);

  MTL::Function *vertexFunction = library->newFunction(vertexName);
  MTL::Function *fragmentFunction = library->newFunction(fragmentName);

  renderDescriptor->setVertexFunction(vertexFunction);
  renderDescriptor->setFragmentFunction(fragmentFunction);
  renderDescriptor->colorAttachments()->object(0)->setPixelFormat(
      MTL::PixelFormatBGRA8Unorm);

  _renderPipelineState =
      _device->newRenderPipelineState(renderDescriptor, &error);

  if (!_renderPipelineState) {
    std::cerr << "Render pipeline error: "
              << (error ? error->localizedDescription()->utf8String()
                        : "unknown")
              << std::endl;
  }

  if (vertexFunction)
    vertexFunction->release();
  if (fragmentFunction)
    fragmentFunction->release();
  renderDescriptor->release();
  library->release();
}

void Renderer::initParticles() {
  NS::UInteger bufferSize = sizeof(Particle) * _particleCount;
  _particleBuffer =
      _device->newBuffer(bufferSize, MTL::ResourceStorageModeShared);
}

void Renderer::resetParticles(float width, float height) {
  if (!_particleBuffer)
    return;

  Particle *particles = (Particle *)_particleBuffer->contents();

  int cols = std::sqrt(_particleCount);
  int rows = _particleCount / cols;

  float spacingX = width / cols;
  float spacingY = height / rows;

  for (int i = 0; i < _particleCount; i++) {
    int col = i % cols;
    int row = i / cols;

    float x = col * spacingX + spacingX * 0.5f;
    float y = row * spacingY + spacingY * 0.5f;

    particles[i].position = {x, y};
    particles[i].velocity = {0, 0};
    particles[i].originalPosition = {x, y};
  }
}

void Renderer::drawableSizeWillChange(float width, float height) {
  _viewportWidth = width;
  _viewportHeight = height;
  resetParticles(width, height);
}

void Renderer::updateMousePosition(float x, float y) {
  _mouseX = x;
  _mouseY = y;
}

void Renderer::draw() {
  NS::AutoreleasePool *pool = NS::AutoreleasePool::alloc()->init();

  if (!_layer) {
    pool->release();
    return;
  }

  CA::MetalDrawable *drawable = _layer->nextDrawable();
  if (!drawable) {
    pool->release();
    return;
  }

  MTL::CommandBuffer *commandBuffer = _commandQueue->commandBuffer();

  // Update Uniforms
  Uniforms uniforms;
  uniforms.mousePosition = {_mouseX, _mouseY};
  uniforms.time = 0.0f; // CACurrentMediaTime() would need CoreFoundation
  uniforms.resolution = {_viewportWidth, _viewportHeight};
  uniforms.repulsionRadius = 300.0f;
  uniforms.repulsionStrength = 2.0f;

  // Compute Pass
  if (_computePipelineState) {
    MTL::ComputeCommandEncoder *computeEncoder =
        commandBuffer->computeCommandEncoder();
    computeEncoder->setComputePipelineState(_computePipelineState);
    computeEncoder->setBuffer(_particleBuffer, 0, 0);
    computeEncoder->setBytes(&uniforms, sizeof(Uniforms), 1);

    NS::UInteger w = _computePipelineState->threadExecutionWidth();
    MTL::Size threadGroupSize = MTL::Size::Make(w, 1, 1);
    MTL::Size threadGroups =
        MTL::Size::Make((_particleCount + w - 1) / w, 1, 1);

    computeEncoder->dispatchThreadgroups(threadGroups, threadGroupSize);
    computeEncoder->endEncoding();
  }

  // Render Pass
  MTL::RenderPassDescriptor *passDescriptor =
      MTL::RenderPassDescriptor::alloc()->init();
  MTL::RenderPassColorAttachmentDescriptor *colorAttachment =
      passDescriptor->colorAttachments()->object(0);
  colorAttachment->setTexture(drawable->texture());
  colorAttachment->setLoadAction(MTL::LoadActionClear);
  colorAttachment->setClearColor(MTL::ClearColor::Make(0.1, 0.1, 0.1, 1.0));
  colorAttachment->setStoreAction(MTL::StoreActionStore);

  MTL::RenderCommandEncoder *renderEncoder =
      commandBuffer->renderCommandEncoder(passDescriptor);
  if (renderEncoder && _renderPipelineState) {
    renderEncoder->setRenderPipelineState(_renderPipelineState);
    renderEncoder->setVertexBuffer(_particleBuffer, 0, 0);
    renderEncoder->setVertexBytes(&uniforms, sizeof(Uniforms), 1);

    renderEncoder->drawPrimitives(MTL::PrimitiveTypePoint, NS::UInteger(0),
                                  NS::UInteger(_particleCount));
    renderEncoder->endEncoding();
  }

  commandBuffer->presentDrawable(drawable);
  commandBuffer->commit();

  passDescriptor->release();
  pool->release();
}
